src/Messenger/IndexationRequestHandler.php

```diff
@@ -26,27 +28,67 @@ abstract class IndexationRequestHandler implements MessageHandlerInterface
 
     private $client;
 
-    public function __construct(Client $client)
+    private $bus;
+
+    public function __construct(Client $client, MessageBusInterface $bus)
     {
         $this->client = $client;
+        $this->bus = $bus;
 
         // Disable the logs for memory concerns
         $this->client->setLogger(new NullLogger());
     }
 
     public function __invoke(IndexationRequestInterface $message)
     {
-        $indexer = $this->client->getIndexer();
-
+        $messages = [];
         if ($message instanceof MultipleIndexationRequest) {
-            foreach ($message->getOperations() as $operation) {
-                $this->schedule($indexer, $operation);
-            }
+            $messages = $message->getOperations();
         } elseif ($message instanceof IndexationRequest) {
-            $this->schedule($indexer, $message);
+            $messages = [$message];
         }
 
-        $indexer->flush();
+        $indexer = $this->client->getIndexer();
+
+        $messageOffset = 0;
+        $responseOffset = $indexer->getQueueSize();
+
+        try {
+            foreach ($messages as $indexationRequest) {
+                ++$messageOffset;
+                $this->schedule($indexer, $indexationRequest);
+
+                if (0 === $indexer->getQueueSize()) {
+                    $responseOffset = 0;
+                }
+            }
+
+            $indexer->flush();
+        } catch (ResponseException $exception) {
+            // Extracts failed operations from the bulk
+            // Responses are checked in reverse mode because we have only requests from the last bulk
+            $failedMessages = [];
+            $allResponses = $exception->getResponseSet()->getBulkResponses();
+            $concernedResponses = array_reverse(array_slice($allResponses, $responseOffset));
+            $executedMessages = array_reverse(array_slice($messages, 0, $messageOffset));
+            foreach ($concernedResponses as $key => $response) {
+                if (!$response->isOk()) {
+                    array_unshift($failedMessages, $executedMessages[$key]);
+                }
+            }
+
+            // Throws exception as-is if all operations have failed
+            if (count($failedMessages) === count($messages)) {
+                throw $exception;
+            }
+
+            // Redispatch failed or non-executed messages
+            $nonExecutedMessages = array_slice($messages, $messageOffset);
+            $toRedispatch = array_merge($failedMessages, $nonExecutedMessages);
```

Comment:
We may want to dispatch `$nonExecutedMessages` as a single `MultipleIndexationRequest` instead

Reply:
done via 5a20a5e16b862bae83f4e79d8e56fb4502c488b6